### ISSUE TITLE
Supporting Selective Activation Checkpointing and CPU Offloading Option.

### ISSUE THREAD

**Issue by SeunghyunSEO** [huggingface/transformers#29648]
### Feature request

- references
  - [deepspeed docs](https://deepspeed.readthedocs.io/en/latest/activation-checkpointing.html)

### Motivation

Hi, first of all, thank you for your great contribution for open-source community, HF team !
i really enjoy training with HF transformers but found that activation checkpointing (a.k.a gradient checkpointing) is applied for every layers and it checkpoint activation in GPU memory and it can easily leads to GPU OOM.
However, there are well-known techniques to checkpoint activations selectively and even in CPU.
Implementations of these features can be found in Megatron and DeepSpeed, and torch team also release [the blog](https://pytorch.org/blog/maximizing-training/#selective-activation-checkpointing) about this technique today.  

So i tested selective activation checkpointing with CPU offloading by modifying HF model (in my case, gpt-2 xl).
i just changed below two `model._gradient_checkpointing_func` and forward pass like below.

```python
from transformers import AutoConfig, AutoModelForCausalLM
from deepspeed.runtime.activation_checkpointing import checkpointing

# init model and activate grad ckpt
model = AutoModelForCausalLM(**kwargs)
model.gradient_checkpointing_enable()

# change grad ckpt function
checkpointing. configure(
    mpu_=None,
    deepspeed_config=None,
    partition_activations=False,
    contiguous_checkpointing=False,
    num_checkpoints=num_checkpoints, # only use these for now
    checkpoint_in_cpu=True, # only use these for now
    synchronize=False,
    profile=True,
)
model._gradient_checkpointing_func = checkpointing.checkpoint # use deepspeed activation checkpointing function
```

```python
    def custom(start, end):
        def custom_forward(*inputs):
            layers_ = self.h[start:end]
            x_ = inputs[0] # hidden_states
            head_mask_ = inputs[3] # hidden_states
            for i, layer in enumerate(layers_):
                x_ = layer(
                    x_, 
                    None,
                    inputs[2],
                    head_mask_[i],
                    inputs[4],
                    inputs[5],
                    False,
                    None
                )[0] # tuple to tensor
                # Tra()
            return x_
        return custom_forward

    if self.gradient_checkpointing and self.training:
        # print('use_deepspeed_activation_checkpointing is True !')
        # Tra()
        l = 0
        total_num_layers = len(self.h)
        chunk_length = num_layers ## how many checkpoint do you want?
        # print(f'chunk_length : {chunk_length}')

        while l < total_num_layers:
            if self.model_parallel:
                raise NotImplementedError("Idc model parallel :)")
            if output_hidden_states:
                raise NotImplementedError("Idc layerwise output :)")

            hidden_states = self._gradient_checkpointing_func(
                custom(l, l+chunk_length),
                hidden_states,
                None,
                attention_mask,
                head_mask[l:l+chunk_length],
                encoder_hidden_states,
                encoder_attention_mask,
                False,
                None,
            )

            if output_attentions:
                raise NotImplementedError("We use efficient SDPA and it does not support to output attention score map")
            if self.model_parallel:
                raise NotImplementedError("Idc model parallel :)")

            l += chunk_length
```

and i could see significant improvement in GPU memory (trading off throughput).

<img width="1238" alt="cpu_gpu_elapsed_from_gpt4" src="https://github.com/huggingface/transformers/assets/48202736/9f9d102c-4c35-4216-b8b7-440fdc127ff8">

The quantitative results (CPU/GPU memory and elapsed time for 3 times forwarding) are recorded by torch profiler and figure was generated by GPT-4. 
My setting was like this.

- 2x A100 80GB
- gpt2-xl (near 1.5B)
- fixed batch size with same random seed (`[128, 512] # B, T`)
- zero-3 without CPU offloading (using accelerate)
- flash attn 2 is applied (monkey patched with xformers)
- fixed input tensor where `Batch_size, Sequence_lenth = 128, 512`

Results show that it could save about 10GB for each GPU device.
The reason why reserved GPU memory for checkpointing every 2, 3, 6 layers are same is that peak memory is dominated by computing loss i guess.
here is my detailed result for checkpointing every 3 layers with CPU offloading,

<img width="1047" alt="every3_cpu" src="https://github.com/huggingface/transformers/assets/48202736/4a9ee4e7-7b7e-4e51-96e7-78868943dda4">

<img width="1047" alt="every3_gpu" src="https://github.com/huggingface/transformers/assets/48202736/df40e4b6-f1fb-4cff-aebb-f85aa0e39c64">

and HF default version (using torch.utils.checkpoint for every layers)

<img width="1057" alt="default_cpu" src="https://github.com/huggingface/transformers/assets/48202736/8455b32e-2410-4faa-b093-b918c1ebe4d0">

<img width="1063" alt="default_gpu" src="https://github.com/huggingface/transformers/assets/48202736/61ab67b6-2d54-4388-8fe1-5f7ebdc06967">

i didn't add profiling result for larger model but there were more improvement and i can find sweet spot for GPU memory VS throughput trade off.  

### Your contribution

Didnt PR yet because every model have their own forward method and applying it for all would be a huge update.
i just add profiling results and naive code snippet for implementation.
just want to discuss with you guys !

**Comment by SeunghyunSEO** [comment]
@amyeroberts hi Amy, i hope you're doing well. i'm wondering if you could help me understand why only this post hasn't received a response yet? ty in advance!

**Comment by amyeroberts** [comment]
Hi @SeunghyunSEO, apologies for the delay in response. Thank you for such a detail and extensive write-up and proposal! 

I'm going to cc in @younesbelkada here, who is the transformers king of optimizations and will have a better idea of the correct approach here. 

In general, we want to try to avoid making our forward passes unnecessarily complex e.g. having lots of if/else statements., so if we want to add this, we'd have to find a solution which doesn't do this. Maybe we can think of abstracting out the gradient checkpointing function, similar to how we do for attention, allowing us to switch between approaches WDYT @younesbelkada ?

**Comment by younesbelkada** [comment]
hi @SeunghyunSEO @amyeroberts 
Thanks so much ! The proposal is really nice, thanks for this detailed issue and the experiments ! 
I think it could make sense to abstract things out the gradient checkpointing function to make it much more modulable, can you open a draft PR with your changes on a single model and we can take it from there to see if we can find a more generic way to do so, what do you think?

**Comment by SeunghyunSEO** [comment]
hi @SeunghyunSEO , i open a draft PR :)
sorry for the late, its been busy days haha...

**Comment by SeunghyunSEO** [comment]
> Hi @SeunghyunSEO sorry for the late reply, will have a look ASAP ! 🙏

its ok my man, @younesbelkada :)
btw i found one more potential issue for using deepspeed grad ckpt function.
if we checkpoint activations selectively in every 4 layers and we want to get layerwise output (`output_hidden_states`), 
some items of `output_hidden_states` become scalar not tensor.
i guess it is because 4th layer's output is fed to next 4 layers and it manipulate 4th layer's output...?

**Comment by zhangvia** [comment]
> its ok my man, [@younesbelkada](https://github.com/younesbelkada) :) btw i found one more potential issue for using deepspeed grad ckpt function. if we checkpoint activations selectively in every 4 layers and we want to get layerwise output (`output_hidden_states`), some items of `output_hidden_states` become scalar not tensor. i guess it is because 4th layer's output is fed to next 4 layers and it manipulate 4th layer's output...?

hey, i encounter the same issue. it seems that the deepspeed would modify the inputs to scalar. something like this:
```python
for block in self.blocks:
    hidden_states, encoder_hiddenstates = block(hidden_states,
                                                encoder_hidden_states,
                                                time_steps,
                                                )
```
if the block is checkpointed, the time_steps will be changed to scalar.

do you find any solution? @SeunghyunSEO

**Comment by aeeeeeep** [comment]
You only need to modify `checkpoint` function in deepspeed/runtime/activation_checkpointing/checkpointing.py like this:

```python
@compiler.disable  # WA from Pytorch repo for compile + zero 3 accuracy issue
def checkpoint(function, *args):
    """Checkpoint a model or part of the model.
    This has been directly copied from torch.utils.checkpoint. """

    all_outputs = []
    return CheckpointFunction.apply(function, all_outputs, *args)
    # if len(all_outputs) == 1:
    #     return all_outputs[0]
    # else:
    #     return tuple(all_outputs)
```

It just a draft.

**Comment by aeeeeeep** [comment]
I have experimented with implementing a relatively general activation offloading mechanism in Transformers based on the approach below, but the performance is 5 times slower compared to recomputation!

This is because in most scenarios, the DP degree during fine-tuning is set very high, resulting in extremely large activations that need to be offloaded. Additionally, offloading can easily saturate the CPU memory. The reference implementation I followed is also only applicable to single-GPU setups. In Megatron-style training, TP and PP are typically used, while DP is kept relatively small. In such cases, activations are distributed across multiple GPUs, and the CPU memory is generally sufficient.

https://github.com/pytorch/torchtune/pull/1443

**Comment by qgallouedec** [comment]
Hey @SeunghyunSEO, sorry for the very slow follow-up here, we're catching up on the backlog, and thank you again for the unusually thorough write-up and profiling.

Good news: both things you asked for are reachable today without any change to transformers. `gradient_checkpointing_enable` forwards `gradient_checkpointing_kwargs` straight into `torch.utils.checkpoint.checkpoint` https://github.com/huggingface/transformers/blob/be49cba7116bcbe4a7a3f10d324e25fef8b17942/src/transformers/modeling_utils.py#L3187-L3204
so selective checkpointing is just `model.gradient_checkpointing_enable(gradient_checkpointing_kwargs={"use_reentrant": False, "context_fn": lambda: create_selective_checkpoint_contexts(policy_fn)})`, and CPU offloading works by swapping `_gradient_checkpointing_func` for one that wraps the call in `saved_tensors_hooks`.

What's left is the every-N-layers chunking, and that's the part we'd rather not take: it means editing the decoder loop in every model file, a new config knob we'd be stuck supporting forever, and it breaks `output_hidden_states`/`output_attentions` (the scalar-output bug you and @lllllinux both hit).
With torch's op-level SAC now covering the same memory/throughput trade-off more finely, and @aeeeeeep's datapoint that offloading came out ~5x slower than recompute in the DP-heavy setups typical of fine-tuning, we don't think the carve-outs pay for themselves. Closing this and the draft PR, but genuinely grateful for the investigation!
